### PR TITLE
Migrate from nose to unittest python test framework

### PR DESCRIPTION
**Pull request by jkonecny12**
Migration is required because nose tests doesn't work on python 3.10 and we should have done this a long time anyway.
Change naming of test, Makefile targets and vars to make everything consistent. Make the naming just `unit` to make it more general to use and avoid such a big renaming in the future.

Unfortunately, I'm disabling glade tests because they are plugin for nose tests so I won't expect it to be an easy migration. (I could be surprised).

Seems that the frameworks have a different coverage output. I'm expecting coverage failure.

### CODE REVIEW

File: tests/README.rst [hunk 1]
```diff
@@ -61,12 +64,15 @@ Executing the tests can be done with::
...
+
 
+To run a subset of unit tests do::
 
-To run a subset of nose tests do::
+    make TESTS=unit_tests.sh UNIT_TESTS_PATTERN='test_layout_variant_' check
 
-    make TESTS=nosetests.sh NOSE_TESTS_ARGS="nosetests/pyanaconda_tests/keyboard_test.py nosetests/pyanaconda_tests/timezone_test.py" check
+The ``UNIT_TESTS_PATTERN`` variable is passed to ``unittest discover -k``. See
+the documentation for more info.
```
Comment (M4rtinK) [inline tests/README.rst:75]: Might be a good idea to link to the respective documentation or mention what documentation to check - man page or some Python module documentation ?

**Review by M4rtinK — APPROVED** [review]
Looks good to me as far as I can tell! Thanks a lot for handling something so big and critical so quickly! :)

File: tests/unit_tests_root.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
+
+# Run only the nosetests that require root
+
+srcdir=${srcdir:=$(dirname "$0")}
+exec ${srcdir}/unit_tests.sh ${srcdir}/unit_tests/pyanaconda_tests/user_create_test.py \
+                             ${srcdir}/unit_tests/pyanaconda_tests/storage_test.py
```
Comment (poncovka) [inline tests/unit_tests_root.sh:7]: I don't think these files exist. Maybe we can remove this script.

**Review by poncovka — APPROVED** [review]
Looks great to me. Thanks!

**Review by rvykydal — APPROVED** [review]
Looks good to me, thank you.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Removed `nosetests_root.sh`.
* Add link to README file.

**Comment by jkonecny12** [comment]
Thanks a lot for your suggestions!

File: tests/Makefile.am [hunk 3]
```diff
@@ -38,34 +39,34 @@ EXTRA_DIST = README.rst usercustomize.py
...
 PEP8_TARGETS ?= ""
 
 # Test scripts need to be listed both here and in TESTS
 dist_check_SCRIPTS = $(srcdir)/glade/*.py \
 		     $(srcdir)/lib/*.py \
 		     $(srcdir)/lib/*.sh \
-		     nosetests.sh \
-		     nosetests_root.sh \
+		     unit_tests.sh \
+		     unit_tests_root.sh \
```
Comment (poncovka) [inline tests/Makefile.am:51]: This should be removed as well.

Comment (jkonecny12) [reply tests/Makefile.am:51]: ahh, good point!

**Comment by jkonecny12** [comment]
**Rebased:**
* Fixed issue introduced by last change.

**Review by poncovka — APPROVED** [review]

File: tests/unit_tests/pyanaconda_tests/test_ks_version.py [hunk 4]
```diff
@@ -59,7 +59,7 @@ def test_data(self):
 
     def test_dracut_commands(self):
         """Test that dracut uses the right versions of kickstart commands"""
-        # top_srcdir should have been set by nosetests.sh. If it wasn't, the KeyError
+        # top_srcdir should have been set by unit_tests.sh. If it wasn't, the KeyError
```
Comment (VladimirSlavik) [inline tests/unit_tests/pyanaconda_tests/test_ks_version.py:62]: This is changed too early - the file is not renamed yet in that commit - but never mind.

File: Makefile.am [hunk 5]
```diff
@@ -329,12 +329,12 @@ tests-pylint:
 	$(MAKE)
 	$(MAKE) TMPDIR=/var/tmp TEST_SUITE_LOG=test-suite.log TESTS=pylint/runpylint.py check
 
-tests-nose-only:
+tests-units-only:
```
Comment (VladimirSlavik) [inline Makefile.am:332]: I have never seen "units" as a plural for tests. Feel free to prove me wrong, though!

File: tests/lib/gladecheck.py [hunk 6]
```diff
@@ -38,6 +38,8 @@
 import os
 import unittest
 import copy
+# FIXME: Enable glade tests back when we found a way how to migrate them to python3 unittest
+# pylint: disable=import-error
```
Comment (VladimirSlavik) [inline tests/lib/gladecheck.py:42]: We could probably convert this to "just" a normal test that runs standalone, and use [`subTest`](https://docs.python.org/3/library/unittest.html?highlight=unittest#distinguishing-test-iterations-using-subtests) to report the failures in a sane way.

File: tests/rpm_tests.sh [hunk 7]
```diff
@@ -1,8 +1,16 @@
...
+fi
+
 # If no tests were selected by user or makefile, select all of them
-if [ $# -eq 0 ] && [ -z $RPM_TESTS_ARGS ]; then
+if [ $# -eq 0 ]; then
     set -- "${top_srcdir}"/tests/rpm_tests
 fi
 
-exec python3 -m nose -v --exclude=logpicker -a \!acceptance,\!slow $RPM_TESTS_ARGS "$@"
+# TODO: RPM_TESTS_ARGS are not usable in the container environment. Do we want to keep it?
```
Comment (VladimirSlavik) [inline tests/rpm_tests.sh:14]: Have we used it for anything so far? I think no?

Comment (jkonecny12) [reply tests/rpm_tests.sh:14]: Yeah, I'm convinced to remove this but I don't want to do it in this PR.

File: tests/unit_tests/pyanaconda_tests/test_ks_version.py (hunk 4 above)
Comment (jkonecny12) [reply tests/unit_tests/pyanaconda_tests/test_ks_version.py:62]: Fixed.

File: Makefile.am (hunk 5 above)
Comment (jkonecny12) [reply Makefile.am:332]: Fixed.

File: tests/unit_tests.sh [hunk 8]
```diff
@@ -7,8 +7,8 @@ if [ -z "$top_srcdir" ]; then
 fi
 
 # If no tests were selected by user or makefile, select all of them
-if [ $# -eq 0 ] && [ -z "$NOSE_TESTS_ARGS" ]; then
+if [ $# -eq 0 ]; then
     set -- $top_srcdir/tests/unit_tests
 fi
 
-exec python3 -m unittest discover -t $top_srcdir -v -b -s $@ $NOSE_TESTS_ARGS
+exec python3 -m unittest discover -t $top_srcdir -v -b ${UNIT_TESTS_PATTERN:+-k} $UNIT_TESTS_PATTERN -s $@
```
[the remark this answers is not in the document]
Comment (jkonecny12) [reply tests/unit_tests.sh:14]: Great improvement!

File: tests/lib/gladecheck.py (hunk 6 above)
Comment (jkonecny12) [reply tests/lib/gladecheck.py:42]: Seems like a good idea but not for this PR.

**Comment by jkonecny12** [comment]
**UPDATED:**
* Thanks @VladimirSlavik for your suggestions.

**Review by VladimirSlavik — APPROVED** [review]
Wow, you didn't have to...

**Review by VladimirSlavik — APPROVED** [review]
Thank you!

**Pull request merged**